crates/ruff_linter/src/rules/pylint/rules/access_member_before_definition.rs

```diff
@@ -0,0 +1,154 @@
+use ast::{call_path::collect_call_path, Stmt};
+use ruff_diagnostics::{Diagnostic, FixAvailability, Violation};
+use ruff_macros::violation;
+use ruff_python_semantic::ScopeKind;
+
+use ruff_text_size::Ranged;
+use rustc_hash::FxHashSet;
+use similar::DiffableStr;
+
+use crate::checkers::ast::Checker;
+use ruff_python_ast::{self as ast, Expr};
+
+/// Checks for accesses on uninitialized class members.
+///
+/// ## Example
+/// ```python
+/// class Unicorn:
+//  def __init__(self, x):
+//     if self.x > 9000:  # [access-member-before-definition]
+//         pass
+//     self.x = x
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class Unicorn:
+//  def __init__(self, x):
+//     self.x = x
+//     if self.x > 9000:
+//         pass
+/// ```
+#[violation]
+pub struct AccessMemberBeforeDefinition {
+    member: String,
+}
+
+impl Violation for AccessMemberBeforeDefinition {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;
+
+    fn message(&self) -> String {
+        let AccessMemberBeforeDefinition { member } = self;
+        format!("Accessed `{member}` before it was initialized",)
+    }
+
+    fn message_formats() -> &'static [&'static str] {
+        todo!()
+    }
+}
+
+pub(crate) fn access_member_before_definition(checker: &mut Checker, expr: &Expr) {
+    let Expr::Attribute(ast::ExprAttribute { value, attr, .. }) = expr else {
+        return;
+    };
+
+    if let Some(call_path) = collect_call_path(value) {
+        // Ignore `cls` accesses.
+        if matches!(call_path.as_slice(), ["cls"]) {
+            return;
+        }
+    }
+
+    // return if we are not in a class
+    if !checker.semantic().current_scope_ids().any(|scope_id| {
+        let scope = checker.semantic().scopes.get(scope_id).unwrap();
+        scope.kind.is_class()
+    }) {
+        return;
+    }
+
+    let init_function = checker.semantic().current_scope_ids().find_map(|scope_id| {
+        let scope = checker.semantic().scopes.get(scope_id).unwrap();
+        match &scope.kind {
+            ScopeKind::Function(function) => {
+                if function.name.as_str() == "__init__" {
+                    Some(function)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    });
+
+    let function_body = match init_function {
+        Some(function) => &function.body,
+        None => return,
+    };
+
+    let mut assigned_members: FxHashSet<&str> = FxHashSet::default();
+    for stmt in function_body {
+        // if the assignment is after the current expression then ignore it
+        if stmt.range().start() > expr.range().start() {
```

Comment:
(Nit: you can do `stmt.start()` and `expr.start()`. Anything that has `.range()` automatically gets `.start()` and `.end()`.)